lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/Task/TaskStatus.cs

```diff
@@ -98,28 +98,48 @@ public void SetException(Exception e)
         {
             lock (_heartBeatManager)
             {
-                try
+                if (_lastException.IsPresent())
                 {
-                    if (!_lastException.IsPresent())
-                    {
-                        _lastException = Optional<Exception>.Of(e);
-                    }
+                    return;
+                }
 
-                    State = TaskState.Failed;
+                _lastException = Optional<Exception>.Of(e);
+                State = TaskState.Failed;
+
+                try
+                {
                     _taskLifeCycle.Stop();
                 }
-                finally
+                catch (Exception stopException)
                 {
-                    Heartbeat();
+                    Utilities.Diagnostics.Exceptions.Caught(
```

Comment:
I must admit that I am somewhat confused about this. Task crash and occur at three times:

  * Before it starts: During construction or when invoking handlers for `TaskStart`.
  * While it runs: During `ITask.Call()`
  * After it is done: During `ITask.Dispose()` or in the handlers of `TaskStop`.

Now, the code at hand actually invokes the `TaskStop` handlers when there was a failure earlier, right? That seems counter-intuitive. Why would one want to call a handler for `TaskStop` on an unsuccessful exit?

The corresponding java code (`org.apache.reef.runtime.common.evaluator.task.TaskStatus.setException()`) does not do so:

```java
  void setException(final Throwable throwable) {
    synchronized (this.heartBeatManager) {
      this.lastException = Optional.of(throwable);
      this.state = State.FAILED;
      this.check();
      this.heartbeat();
    }
  }
```

Further, the method which runs the Task, `org.apache.reef.runtime.common.evaluator.task.TaskRuntime.run()`, makes the state in which the Task was during the exception known to the Driver:

```java
  public void run() {
    try {
      // Change state and inform the Driver
      this.taskLifeCycleHandlers.beforeTaskStart();

      LOG.log(Level.FINEST, "Informing registered EventHandler<TaskStart>.");
      this.currentStatus.setRunning();

      // Call Task.call()
      final byte[] result = this.runTask();

      // Inform the Driver about it
      this.currentStatus.setResult(result);

      LOG.log(Level.FINEST, "Informing registered EventHandler<TaskStop>.");
      this.taskLifeCycleHandlers.afterTaskExit();

    } catch (final TaskStartHandlerFailure taskStartHandlerFailure) {
      LOG.log(Level.WARNING, "Caught an exception during TaskStart handler execution.", taskStartHandlerFailure);
      this.currentStatus.setException(taskStartHandlerFailure.getCause());
    } catch (final TaskStopHandlerFailure taskStopHandlerFailure) {
      LOG.log(Level.WARNING, "Caught an exception during TaskStop handler execution.", taskStopHandlerFailure);
      this.currentStatus.setException(taskStopHandlerFailure.getCause());
    } catch (final TaskCallFailure e) {
      LOG.log(Level.WARNING, "Caught an exception during Task.call().", e.getCause());
      this.currentStatus.setException(e);
    }
  }
```

That way, the Driver can initiate the cleanup, if one is necessary after the Task failed. I haven't checked, but hope we don't change those exceptions on the way to the Driver.